File: src/storeFx/initializePlayQueue.ts
import * as Fx from '../fx/fx';
import { mergeIntoMap } from '../mergeIntoMap/mergeIntoMap.js';
import type { Store } from '../store/types';
import { getPlayQueue } from '../subsonic';
import { setCurrentTime } from './setCurrentTime';

export function initializePlayQueue() {
  return Fx.bind(Fx.ask<{ store: Store }>(), ({ store }) => {
    const credentials = store.getState().auth.credentials;
    if (!credentials) return Fx.Ok();

    return Fx.bind(
      Fx.catch_(Fx.provide(getPlayQueue(), { credentials }), error =>
        error.kind === 'no-credentials' ||
        error.kind === 'network-error' ||
        error.kind !== 'api-error' ||
        error.code !== 70
          ? Fx.Err(error)
          : Fx.Ok(undefined),
      ),
      playQueue => {
        const currentSongId =
          playQueue?.current != null ? String(playQueue.current) : undefined;

        if (currentSongId != null) {
          Fx.runAsync(
            setCurrentTime(() => (playQueue?.position ?? 0) / 1000),
            { store },
          );
        }

        store.setState(prevState => {
          const state = store.getState();

          return {
            player: {
              ...prevState.player,
              currentSongId,
              queuedSongIds: playQueue?.entry.map(song => song.id) ?? [],
            },
            songs: {
              byId:
                playQueue == null
                  ? state.songs.byId
                  : mergeIntoMap(state.songs.byId, playQueue.entry, x => x.id),
            },
          };
        });

        return Fx.Ok();
      },
    );
  });
}

File: src/storeFx/initializePlayQueue.ml
open Fx
open Monad_syntax

let make () =
  let* Deps.{ store } = ask () in
  let state = store |. Zustand.getState in
  match state.auth.credentials with
  | None -> Ok ()
  | Some credentials ->
      let* playQueue =
        Subsonic.GetPlayQueue.make ()
        |. provide ~deps:{ credentials = Some credentials }
        |. catch (fun error ->
            match error with
            | ApiError { code; _ } when code == 70 -> Ok None
            | error -> Error error)
      in
      let open Subsonic.GetPlayQueue.Response in
      let currentSongId = playQueue |. Option.bind (fun x -> x.current) in
      currentSongId
      |> Option.iter (fun _ ->
          SetCurrentTime.make (fun _ ->
              (playQueue
              |. Option.bind (fun x -> x.position)
              |> Option.value ~default:0.0)
              /. 1000.0)
          |. runAsync ~deps:{ store }
          |> ignore);
      Ok
        (store
        |. Zustand.setState (fun prevState ->
            {
              prevState with
              player =
                {
                  prevState.player with
                  currentSongId;
                  queuedSongIds =
                    playQueue
                    |> Option.map (fun x ->
                        x.entry
                        |. Js.Array.map ~f:(fun song -> Subsonic.Song.(song.id)))
                    |> Option.value ~default:[||];
                };
              songs =
                {
                  byId =
                    (let state = store |. Zustand.getState in
                     match playQueue |> Option.map (fun x -> x.entry) with
                     | Some entry ->
                         MergeIntoMap.mergeIntoMap state.songs.byId entry
                           (fun x -> x.id)
                     | None -> state.songs.byId);
                };
            }))
